src/mapping/query/function/contains.rs

```diff
@@ -0,0 +1,168 @@
+use super::prelude::*;
+
+#[derive(Debug)]
+pub(in crate::mapping) struct ContainsFn {
+    query: Box<dyn Function>,
+    substring: Box<dyn Function>,
+    case_sensitive: Option<Box<dyn Function>>,
+}
+
+impl ContainsFn {
+    #[cfg(test)]
+    pub(in crate::mapping) fn new(
+        query: Box<dyn Function>,
+        substring: &str,
+        case_sensitive: bool,
+    ) -> Self {
+        let substring = Box::new(Literal::from(Value::from(substring)));
+        let case_sensitive = Some(Box::new(Literal::from(Value::from(case_sensitive))) as _);
+
+        Self {
+            query,
+            substring,
+            case_sensitive,
+        }
+    }
+}
+
+impl Function for ContainsFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let substring = match self.substring.execute(ctx)? {
+            Value::Bytes(b) => b,
+            v => unexpected_type!(v),
+        };
+
+        let value = match self.query.execute(ctx)? {
+            Value::Bytes(b) => b,
+            v => unexpected_type!(v),
+        };
+
+        let mut contains = value
+            .windows(substring.len())
+            .any(|window| window == substring);
+
+        // if we don't find a match, but we are matching with case sensitivity
+        // disabled, we'll have to go the more expensive route of lowercasing
+        // the value and substring.
+        if !contains {
+            let case_sensitive = match &self.case_sensitive {
+                Some(case_sensitive) => match case_sensitive.execute(ctx)? {
+                    Value::Boolean(v) => v,
+                    v => unexpected_type!(v),
+                },
+                None => false,
+            };
+
+            if !case_sensitive {
+                let value = String::from_utf8_lossy(&value).to_lowercase();
+                let substring = String::from_utf8_lossy(&substring).to_lowercase();
+
+                contains = value.contains(&substring);
+            }
+        }
```

Comment:
Consider (only because I'm biased against `let mut`):
```suggestion
        let mut contains = value
            .windows(substring.len())
            .any(|window| window == substring) || {

        // if we don't find a match, but we are matching with case sensitivity
        // disabled, we'll have to go the more expensive route of lowercasing
        // the value and substring.
            let case_sensitive = match &self.case_sensitive {
                Some(case_sensitive) => match case_sensitive.execute(ctx)? {
                    Value::Boolean(v) => v,
                    v => unexpected_type!(v),
                },
                None => false,
            };

            case_sensitive && {
                let value = String::from_utf8_lossy(&value).to_lowercase();
                let substring = String::from_utf8_lossy(&substring).to_lowercase();

                value.contains(&substring)
            }
        }
```

Reply:
Took a slightly (non-mut) different approach, but I love these kinds of improvement suggestions, keep 'm coming 👌

Reply:
I also learned from this :heart: